### DISCUSSION TITLE
Is it possible to use a custom asgi/wsgi server instead of uvicorn?

### DISCUSSION THREAD

**Discussion by itsmeadarsh2008** [zauberzeug/nicegui#3051] [Q&A]
I want to use Granian for the ASGI / WSGI server but don't know how to do it. If someone could help me by just writing the steps out, I shall be very grateful to him/her and I will file a PR.
I searched for the "uvicorn" keyword on the entire repo 
[QUERY](https://github.com/search?q=repo%3Azauberzeug%2Fnicegui+uvicorn&type=code)

**Comment by itsmeadarsh2008** [comment]
Do we need to edit the [server.py](https://github.com/zauberzeug/nicegui/blob/main/nicegui/server.py)?

**Comment by hasansezertasan** [comment]
AFAIK Granian has support for ASGI.

If you get successful, please share with us.

**Reply by itsmeadarsh2008** [reply]
Ok

**Reply by itsmeadarsh2008** [reply]
The tool is good but it lacks programmatic documentation and needs to be more mature with the community. So, it will be a better idea if we let it develop docs for some time.

**Reply by hasansezertasan** [reply]
I see, thank you for sharing your experience using granian with us.

**Reply by gi0baro** [reply]
> The tool is good but it lacks programmatic documentation and needs to be more mature with the community. So, it will be a better idea if we let it develop docs for some time.

Disclaimer: Granian maintainer here.
Which kind of documentation and/or *maturity* are you looking for? Maybe you can open up issues/discussions in Granian repo about those expectations/desires? :)

**Reply by itsmeadarsh2008** [reply]
> > The tool is good but it lacks programmatic documentation and needs to be more mature with the community. So, it will be a better idea if we let it develop docs for some time.
> 
> Disclaimer: Granian maintainer here. Which kind of documentation and/or _maturity_ are you looking for? Maybe you can open up issues/discussions in Granian repo about those expectations/desires? :)

I mean how to use Granian programmatically in Python? Since the uvicorn has programmatical usage docs with Python like `import granian # This kind of documentation`. There is no documentation I have searched so far except a discussion where you have written some of the code. I don't know how to use it in the code directly. As you know, uvicorn is widely implemented. This is the same with the Reflex framework too. They are looking for replacements to uvicorn. However, they have support for external HTTP servers but you have to spawn a new shell for that. Granian is well-suited to performance and can boost API performance. It would be easier for others to integrate Granian in Python esp. Python Web Frameworks like NiceGUI and Reflex. So, the projects can migrate to Granian easily.

**I have doubt:** Is Granian a drop-in replacement for uvicorn?

**Reply by gi0baro** [reply]
> **I have doubt:** Is Granian a drop-in replacement for uvicorn?

It really depends on what do you mean by *drop-in* replacement.
Granian is an ASGI server like Uvicorn, so it speaks the same protocol. CLI speaking, replacing `uvicorn ...` with `granian --interface asgi ...` should work, but if you're looking at the same *programmatical usage* then probably is not just a swap; the Python API for configuring and launching the server is different between the two.

**Reply by itsmeadarsh2008** [reply]
Do you have any docs for granian on configuring it programmatically in Python?? It'll be beneficial for the community. 😄

**Reply by gi0baro** [reply]
> Do you have any docs for granian on configuring it programmatically in Python?? It'll be beneficial for the community. 😄

Gonna add that in the following days/weeks.

**Comment by falkoschindler** [comment]
Isn't this a duplicate of #3039? 🤔

**Reply by hasansezertasan** [reply]
I'm not sure but #3039 recommends switching from uvicorn to granian internally and this discussion is sort of a question about how to use granian with NiceGUI.

Granian supports ASGI/3, WSGI, and RSGI. So I think running a FastAPI (NiceGUI is using FastAPI under the hood) application using granian shouldn't be so hard.

**Reply by itsmeadarsh2008** [reply]
I'm not able to know where to start. We need to remove the uvicorn module from server.py and import granian there and modify the code structure. However, the granian is under active development and doesn't have proper Pythonic documentation which halted the issue. I have sent an issue regarding migration but no one has replied.

**Comment by rodja** [comment]
A first step in this direction would be a demonstration of FastAPI running on a Granian server -- without any NiceGUI code. Is that possible? This would be a necessary precondition for any further steps.

**Reply by itsmeadarsh2008** [reply]
# Granian Versus Uvicorn
> This benchmark is performed on a default blank GitHub Codespace and uses Apache [`ab`](https://httpd.apache.org/docs/2.4/programs/ab.html) for testing. If you're testing it, make sure you have updated the tools using apt and install necessary packages `sudo apt-get install apache2-utils ` under a python virtual environment (`environment`) using `virtualenv`

This is `benchmark.sh`
```bash
#!/bin/bash

# Define the number of requests, concurrency, and duration for the benchmark
NUM_REQUESTS=1000
CONCURRENCY=10
DURATION=45s

# Define the host and ports for uvicorn and granian
UVICORN_PORT=8081
GRANIAN_PORT=8080

echo "Benchmarking Uvicorn Server"
ab -n $NUM_REQUESTS -c $CONCURRENCY -t $DURATION http://localhost:$UVICORN_PORT/ > uvicorn_benchmark.txt

echo "Benchmarking Granian Server"
ab -n $NUM_REQUESTS -c $CONCURRENCY -t $DURATION http://localhost:$GRANIAN_PORT/ > granian_benchmark.txt

echo "Benchmarking completed."
```
Making it an executable
```bash
chmod +x benchmark.sh
```
# Results (independent files: `granian_benchmark.txt` and `uvicorn_benchmark.txt`)
`granian_benchmark.txt`
```txt
This is ApacheBench, Version 2.3 <$Revision: 1843412 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)


Server Software:        granian
Server Hostname:        localhost
Server Port:            8080

Document Path:          /
Document Length:        17 bytes

Concurrency Level:      10
Time taken for tests:   25.275 seconds
Complete requests:      50000
Failed requests:        0
Total transferred:      7100000 bytes
HTML transferred:       850000 bytes
Requests per second:    1978.23 [#/sec] (mean)
Time per request:       5.055 [ms] (mean)
Time per request:       0.506 [ms] (mean, across all concurrent requests)
Transfer rate:          274.32 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.4      0      36
Processing:     0    5   3.8      4      85
Waiting:        0    5   3.8      4      85
Total:          1    5   3.9      4      85

Percentage of the requests served within a certain time (ms)
  50%      4
  66%      4
  75%      4
  80%      5
  90%      8
  95%     11
  98%     16
  99%     22
 100%     85 (longest request)
```
`uvicorn_benchmark.txt`
```txt
This is ApacheBench, Version 2.3 <$Revision: 1843412 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)


Server Software:        uvicorn
Server Hostname:        localhost
Server Port:            8081

Document Path:          /
Document Length:        17 bytes

Concurrency Level:      10
Time taken for tests:   45.002 seconds
Complete requests:      45442
Failed requests:        0
Total transferred:      6452764 bytes
HTML transferred:       772514 bytes
Requests per second:    1009.79 [#/sec] (mean)
Time per request:       9.903 [ms] (mean)
Time per request:       0.990 [ms] (mean, across all concurrent requests)
Transfer rate:          140.03 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.3      0      15
Processing:     1   10  73.5      7    3751
Waiting:        0   10  73.5      6    3751
Total:          1   10  73.6      7    3751

Percentage of the requests served within a certain time (ms)
  50%      7
  66%      7
  75%      8
  80%      9
  90%     11
  95%     14
  98%     20
  99%     25
 100%   3751 (longest request)
```
### Granian is probably faster than so called lightning-fast uvicorn
# `main.py`
```python
from fastapi import FastAPI

app = FastAPI()

@app.get("/")
async def read_root():
    return {"Hello": "World"}

@app.get("/items/{item_id}")
async def read_item(item_id: int, q: str = None):
    return {"item_id": item_id, "q": q}
```

# Commands
```bash
pip install granian uvicorn
granian --interface asgi main:app --port 8080 # Separate Shell
uvicorn main:app --port 8081 # Separate Shell
```

**Comment by itsmeadarsh2008** [comment]
Source Code Soon Will Be Available At This [Repository](https://github.com/itsmeadarsh2008/granian-vs-uvicorn)

**Comment by itsmeadarsh2008** [comment]
Any Updates.

**Comment by rodja** [comment]
I changed this to a Q&A because it's not really formulated as a feature request. Hope someone from the community can help to answer this one.
@itsmeadarsh2008 could you please explain what the difference between this question and the one in #3039 is? We would really like to consolidate the discussion as much as possible.

**Reply by itsmeadarsh2008** [reply]
The difference between #3051 and #3039 is that #3051 explains about using external ASGI as a plugin type (if someone wants to use an HTTP Server other than uvicorn, he/she may able to do so efficiently) and #3039 explains if we could add Granian (A high-speed HTTP server) into NiceGUI codebase to improve performance so that it won't be an issue for slowness and the user doesn't have to look for an alternative. For me, #3051 seems to be the right option because it gives a facility to customize the HTTP Server to fit user needs. If the HTTP Server is heavy for the user (memory effectiveness), he/she would be able to change it to a lighter one.

**Reply by itsmeadarsh2008** [reply]
However, the slowness may not be an issue for us but maybe the user who is deploying to a computer (compute engine) that has limited resources or on a micro can cause performance issues and may crash on high requests.

**Comment by itsmeadarsh2008** [comment]
Updates?

**Reply by rodja** [reply]
No. Someone needs to pick up the topic and experiment with the code. Who would like to give it a try?
